client/src/components/Sidebar/Sidebar.tsx

```diff
@@ -0,0 +1,139 @@
+
+import * as React from 'react';
+
+import Icon from './common/Icon';
+
+export interface Strings {
+  DASHBOARD: string;
+  EDIT_YOUR_ACCOUNT: string,
+  SEARCH: string,
+}
+
+export interface ModuleRenderContext {
+  slim: boolean;
+  expandingOrCollapsing: boolean;
+  currentPath: string;
+  strings: Strings;
+  navigate(url: string): Promise<void>;
+}
+
+export interface ModuleDefinition {
+  render(context: ModuleRenderContext): React.ReactFragment;
+}
+
+export interface SidebarProps {
+  modules: ModuleDefinition[];
+  currentPath: string;
+  strings: Strings;
+  navigate(url: string): Promise<void>;
+  onExpandCollapse?(collapsed: boolean);
+}
+
+export const Sidebar: React.FunctionComponent<SidebarProps> = (
+  { modules, currentPath, strings, navigate, onExpandCollapse }) => {
+  // 'collapsed' is a persistent state that is controlled by the arrow icon at the top
+  // It records the user's general preference for a collapsed/uncollapsed menu
+  // This is just a hint though, and we may still collapse the menu if the screen is too small
+  // Also, we may display the full menu temporarily in collapsed mode (see 'peeking' below)
+  const [collapsed, setCollapsed] = React.useState(window.innerWidth < 800);
+
+  // Call onExpandCollapse(true) if menu is initialised in collapsed state
+  React.useEffect(() => {
+    if (collapsed && onExpandCollapse) {
+      onExpandCollapse(true);
+    }
+  }, []);
+
+  // 'peeking' is a temporary state to allow the user to peek in the menu while it is collapsed, or hidden.
+  // When peeking is true, the menu renders as if it's not collapsed, but as an overlay instead of occupying
+  // space next to the content
+  const [peeking, setPeeking] = React.useState(false);
+
+  // Whether or not to display the menu with slim layout.
+  // Separate from 'collapsed' as the menu can still be displayed with an expanded
+  // layout while in 'collapsed' mode if the user is 'peeking' into it (see above)
+  const slim = collapsed && !peeking;
+
+  // 'expandingOrCollapsing' is set to true whilst the the menu is transitioning between slim and expanded layouts
+  const [expandingOrCollapsing, setExpandingOrCollapsing] = React.useState(false);
+  React.useEffect(() => {
+    setExpandingOrCollapsing(true);
+    const finishTimeout = setTimeout(() => {
+      setExpandingOrCollapsing(false);
+    }, 300);
+
+    return () => {
+      clearTimeout(finishTimeout);
+    };
+  }, [slim]);
+
+  const onClickCollapseToggle = (e: React.MouseEvent) => {
+    e.preventDefault();
+    setCollapsed(!collapsed);
+
+    // Unpeek if the user has just collapsed the menu
+    // Otherwise the menu would just stay open until the mouse leaves
+    if (!collapsed) {
+      setPeeking(false);
+    }
+
+    if (onExpandCollapse) {
+      onExpandCollapse(!collapsed);
+    }
+  };
+
+  // Set up an event handler to switch peeking on/off when the mouse cursor hovers the sidebar
+  const wrapperRef = React.useRef<HTMLDivElement | null>(null);
```

Comment:
Is there a reason you want to use a ref and DOM events rather than adding eg `onMouseEnter` directly to your menu react component below?

Reply:
No reason just didn't think of doing that! I'll change this, thanks!